Managing mandatory service restarts on chef runs

Itxaka opened the pull request:
Allow end users to set a barclamp as "no service reboots allowed" to avoid automatic service restarts by chef .

There is several parts to this:

 - A new experimental config file to enable/disable experimental features like this one.
   - Found at crowbar_framework/config/experimental.yml
   - Gets loaded into the rails config at crowbar start, available at `Rails.application.config.experimental`


 - A chef dsl service resource monkey-patching
   - allows us to skip the service restart in case that a flag is set on the `crowbar-config -> disallow_restarts` databag
   - allows us to set the flag `node[cookbook_name][:requires_restart]` if a restart was skipped due to the `disallow_restart` flag with the barclamp, service name that requires restart, timestamp, and identification if this is a pacemaker service.


 - Several new routes on chef framework:
   - /restart_management/configuration
     - GET: List the barclamps and if service reboots are disallowed for them
     - POST(params: disallow_restarts, barclamp): Sets the disallow_restart flag for a barclamp
    - /restart_management/restarts
      - GET: Lists all of the services needing restarts
      - POST(params: node, service): Cleans the restart flag for the given service in the given node

Itxaka commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Spike chef re-starting services](https://trello.com/c/QKMA5dJu/56-spike-chef-re-starting-services)

Itxaka commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Neutron: Avoid Chef re-starting services](https://trello.com/c/BwtWivrg/67-neutron-avoid-chef-re-starting-services)

rsalevsky commented on crowbar_framework/config/routes.rb:
```diff
@@ -118,6 +118,18 @@
     end
 
+    # service restart management
+    resource :restart_management,
```
Please use the new API v2 namespace.

Itxaka replied:
You gotta be more explicit than that :D

Im guessing you mean setting the routes under the v2 namespace like `/utils/2.0/restart_management` ? Or a different one than that?

We dont have the other util routes under `1.0` which is bad :(

Itxaka replied:
Or you mean to put everything under the `/api/` namespace?

rsalevsky replied:
Somewhere below the following lines would be nice.
```
  namespace :api,
    constraints: ApiConstraint.new(2.0) do
```

rsalevsky commented:
Codewise LGTM but I didn't test it.

matelakat commented on crowbar_framework/spec/controllers/api/restart_management_controller_spec.rb:
```diff
@@ -0,0 +1,209 @@
+      expect(@testing_node["nova"][:disallow_restart]).to be(false)
+    end
+
+    it "alternate calls to enable/disable set the barclamp disallow restart status properly" do
```
I don't get why do we need this test. We already tested the status what happens when we set it to `true` then what happens if we set it to `false`. What are we testing by setting it `true`, `false`, `true` ? Also the comments are misleading, as they are out of sync.

Itxaka replied:
well, we are only testing the transition from a know mocked state into a desired state with the other tests, so I think its useful to test the toggling of the value to check that it works as expected when the mocked state has gone away and we only have real values in there. Plus it doesnt hurt to have more tests.

matelakat commented on crowbar_framework/spec/controllers/api/restart_management_controller_spec.rb:
```diff
@@ -0,0 +1,209 @@
+    end
+
+    it "returns a 404 error if we pass a non-existant barclamp" do
+      post "/api/restart_management/sync", { disallow_restart: true, barclamp: "parrot" }, headers
```
parrot :-)

matelakat commented on crowbar_framework/spec/controllers/api/restart_management_controller_spec.rb:
```diff
@@ -0,0 +1,209 @@
+      time = Time.now.getutc.to_s
+      @admin_node.set["nova"]["requires_restart"] = {
+        "openstack-nova-api" => {
+          "pacemaker_service" => false,
```
Maybe setting one of the `pacemaker_service` datas to `true` would cover that part as well? Also I think then 2 test vectors should be enough?

Itxaka replied:
Well, in here the outcome of the test is more to see that the output is as we expect, more than to check the proper values, especially as the `pacemaker_service` and `timestamp` values are set on the chef runs and we have no way of testing them, so those attributes could be empty for all we care. So only testing the output rather than the actual values or anything like that.

Itxaka commented on chef/cookbooks/utils/libraries/service.rb:
```diff
@@ -0,0 +1,73 @@
+        # to avoid restarts :(
+        if cookbook_name == "apache2"
+          disallow_restart = [
+            node.fetch("horizon".to_s, {}).fetch(:disallow_restart, false),
```
umm, seems like a leftover, let me fix it

Itxaka commented on chef/cookbooks/utils/libraries/service.rb:
```diff
@@ -0,0 +1,73 @@
+          )
+        else
+          # from this point this is a copy of the default method
+          converge_by("restart service #{@new_resource}") do
```
from `Chef::Provider::Service#action_restart` https://github.com/SUSE-Cloud/chef/blob/10-stable-suse/chef/lib/chef/provider/service.rb#L113

Itxaka replied:
I tired to but it seems impossible to me. I may not know enough ruby as to know how to do it, but I expected a super to work here, but that doesnt work, it claims the parent class has no disallow_restart method or something like that.

As the Chef::Provider::Service then gets overload with the specific Service class (InitD, SystemD, etc..) there maybe something special to do here that Im missing. We should get a ruby expert to hlep here methinks.

aspiers replied:
I agree with @matelakat that we should avoid copy'n'paste if we possibly can.  I don't think you can use inheritance because Chef expects to use this exact provider class.  However can't you use `Module#prepend` as described in that StackOverflow answer?  It looks like a clean and simple solution.

matelakat replied:
@aspiers it turned out not to be a copy-paste actually, so the comment was modified to say "this is a modified version of the original method" also asked @rsalevsky on that prepend technique and the decision was to keep it as it is for now.

Itxaka commented:
@matelakat I like that `configuration` name, let me rewrite this to use that as an endpoint instead, because it sounds much better (sometimes I really struggle to come up with proper names, as shown here)

>  The only thing I am thinking about is the use and validation of timestamps, but that could be some further improvements. What I'm looking at is the following scenario:

> Given barclamp has the norestart set
> Joe applies the barclamp at 0400
> Then he restarts services manually at 0415
> Then Joe applies the barclamp again at 0430
> Then Joe posts the re-starts related to his 0415 actions
> In this scenario the system will blindly accept Joe's request to reset the restart requests, whereas they happened ad 0415 and we have done another run since then which might require another restart.
> I think if we added the timestamp to the post data, then we can say: we only delete a record if that happened BEFORE the posted restore timestamp. Again, food for thought.

I would say, lets keep the timestamp but dont do anything with it for the moment. Dealing with timestamp and so on would complicate this, plus it would require another sync to talk about how to manage them, what to do when there is a restart already with an older timestamp, etc..

Lets go for the simpler, required feature from SAP and try to deliver it ASAP, then lets see if they want timestamps or something else and talk about that.

toabctl commented on chef/cookbooks/utils/libraries/service.rb:
```diff
@@ -0,0 +1,73 @@
+      def action_restart
+        # how do we know if its allowed to restart? We use the cookbook name as the key
+
+        # as both keystone and horizon use apache2, they trigger a restart of it when something
```
It's not only horizon and keystone - also barbican, ceilometer, aodh and maybe others in the future

Itxaka replied:
welp, then that screws my whole plan of making this simple :(

Itxaka replied:
Actually in that case we **only** do reloads for the apache service which is ok.

same with horizon, we no longer use any restart for apache, only reloads, and that should be ok i guess?

toabctl commented on chef/cookbooks/utils/libraries/service.rb:
```diff
@@ -0,0 +1,73 @@
+        if cookbook_name == "apache2"
+          disallow_restart = [
+            node.fetch("horizon".to_s, {}).fetch(:disallow_restart, false),
+            node.fetch("keystone", {}).fetch(:disallow_restart, false)
```
not enough. see previous comment.

Itxaka commented:
After several tests with everything possible the only solution I can find its to modify the cookbook_name for the apache2 service during runtime if the restart is triggered like this:

```
diff --git a/chef/cookbooks/keystone/recipes/server.rb b/chef/cookbooks/keystone/recipes/server.rb
index e8e4d1bf6..eafb29945 100644
--- a/chef/cookbooks/keystone/recipes/server.rb
+++ b/chef/cookbooks/keystone/recipes/server.rb
@@ -253,6 +253,28 @@ keystone_register "update keystone endpoint" do
   end
 end
 
+if node[:keystone][:frontend] == "apache"
+  # Needed to set the restart resource for apache as coming from the keystone cookbook
+  # in order to be able to stop keystone rebooting if the flag to disallow_reboot is set
+  # if we didnt do this, the cookbook_name attached to the service restart would be apache2
+  # so we would not be able to differentiate it from other apache2 restarts
+  ruby_block "set apache2 service cookbook_name as keystone" do
+    block do
+      r = resources(service: "apache2")
+      r.cookbook_name = "keystone"
+    end
+    action :nothing
+  end
+
+  ruby_block "set apache2 service cookbook_name back to apache2" do
+    block do
+      r = resources(service: "apache2")
+      r.cookbook_name = "apache2"
+    end
+    action :nothing
+  end
+end
+
 template node[:keystone][:config_file] do
     source "keystone.conf.erb"
     owner "root"
@@ -278,7 +300,13 @@ template node[:keystone][:config_file] do
       rabbit_settings: fetch_rabbitmq_settings
     )
     if node[:keystone][:frontend] == "apache"
+      notifies :create, resources(
+          ruby_block: "set apache2 service cookbook_name as keystone"
+      ), :immediately
       notifies :restart, resources(service: "apache2"), :immediately
+      notifies :create, resources(
+          ruby_block: "set apache2 service cookbook_name back to apache2"
+      ), :immediately
     elsif node[:keystone][:frontend] == "uwsgi"
       notifies :restart, resources(service: "keystone-uwsgi"), :immediately
     end
```

@vuntz @toabctl @dirkmueller @matelakat what do you think about this approach? Its a bit messy but its the best possible way that I have found.

Itxaka commented:
Better solution at https://github.com/crowbar/crowbar-openstack/pull/1244 for the keystone issue.

vuntz commented on chef/cookbooks/utils/libraries/service.rb:
```diff
@@ -0,0 +1,70 @@
+  class Provider
+    class Service < Chef::Provider
+      def action_restart
+        # how do we know if its allowed to restart? We use the cookbook name as the key
```
s/its/it's/

Itxaka commented:
@vuntz unfortunately, that makes too much sense as I did not account for adding new nodes :/

Let me rework this to use databags instead.

Itxaka commented:
@vuntz reworked

Itxaka commented:
wait, no. Not ready. Seems like the databag returned from `::Chef::DataBagItem.load` is in a terrible format

Itxaka commented:
@vuntz now fixed and tested, seems to work as expected :D

vuntz commented:
Asking out loud: instead of `node[cookbook_name][:requires_restart]`, I somehow feel that `node[:crowbar_wall][:requires_restart][cookbook_name]` makes more sense. Is that insane?

Itxaka commented:
@vuntz only insane in that I would need to redo the Pr to account for the change :rage1: . Under `cookbook` seems more self contained to me but other than a vague feeling, I dont really care where it goes :)

vuntz commented on crowbar_framework/app/controllers/api/restart_management_controller.rb:
```diff
@@ -0,0 +1,159 @@
+    barclamp = params[:barclamp]
+
+    # validate that the barclamp exists
+    raise Crowbar::Error::NotFound unless BarclampCatalog.members("openstack").include? barclamp
```
Any reason to restrict that to openstack barclamps?

Itxaka replied:
The card criteria was focused on openstack services (actually just on neutron first, then nova) and it look easier to me to first focus on the openstack services. I mean, do we want to avoid restarts of the dns or network services? Not counting that, for example the network barclamp does a lot of system managing outside of chef resources.

vuntz replied:
I do think there's no reason to restrict this to openstack barclamps, but it's probably okay-ish. No reason to block on this.

vuntz commented on crowbar_framework/app/controllers/api/restart_management_controller.rb:
```diff
@@ -0,0 +1,159 @@
+  def configuration_get
+    # If its a GET call, just show a list of the barclamps disallow_flag status
+    @final_list = Crowbar::DataBagConfig.get_or_create_databag_item(
+      "crowbar-config", "disallow_restart"
```
No opinion yet, but asking: is it fine to mix that in `crowbar-config`, which is where we cache the proposal data?

Itxaka replied:
Didnt want to create an extra databag for that. crowbar-config stores the proposal data on a databag item for `openstack` so this should not be a big issue. <sup><sup>Plus Im very bad at naming things</sup></sup>

vuntz replied:
I guess I can live with it. Not a big fan of the underscore in `disallow_restart`, but change it only if you think it's better with a dash.

aspiers replied:
It would be nice if the use of `_` vs. `-` was decided based on trying to be consistent with whatever's already there.  I haven't checked which that would be in this particular case, but please check if you didn't already.

aspiers replied:
@Itxaka Did you check this?

vuntz commented on crowbar_framework/app/controllers/api/restart_management_controller.rb:
```diff
@@ -0,0 +1,159 @@
+  end
+
+  # get the node names or the cluster node names
+  def get_node_names(elements)
```
This seems unused now.

Itxaka replied:
indeed, removing!

vuntz commented:
> Asking out loud: instead of node[cookbook_name][:requires_restart], I somehow feel that node[:crowbar_wall][:requires_restart][cookbook_name] makes more sense. Is that insane?

After reviewing the patch again, I kept the same feeling and would recommend changing this. Not a deal-breaker, so if you care strongly, I won't object...

vuntz commented:
Also, looking at https://github.com/crowbar/crowbar-ha/pull/224, I think there's some code that could be put in a small library, in order to avoid having crowbar-ha know some internals...

Itxaka commented:
> After reviewing the patch again, I kept the same feeling and would recommend changing this. Not a deal-breaker, so if you care strongly, I won't object...

changed ¬_¬


> Also, looking at crowbar/crowbar-ha#224, I think there's some code that could be put in a small library, in order to avoid having crowbar-ha know some internals...

I thougth about that but IMO it would be way too verbose, as we would need like 3 different, really small, functions (maybe in the CrowbarHelper ..... helper). TBH I rather leave this so its more clear what is doing rather than obfuscating it more behind smaller functions in a different part. This is already monkeypatching a chef resource which is not an ideal situation, so I would like it to be as easy to see as possible (thus all the long comments in there)

vuntz commented on crowbar_framework/app/controllers/api/restart_management_controller.rb:
```diff
@@ -0,0 +1,147 @@
+    node = NodeObject.find("name:#{node_name} AND requires_restart:#{service}").first
+
+    # validate that node exists
+    raise Crowbar::Error::NotFound if node.nil?
```
I don't think returning `NotFound` is good here, due to the query itself: the node may be found, but with no `requires_restart` flag. I'd somehow prefer if we return `NotFound` if the node doesn't exist, and something else (or even no error at all!) of there's no `requires_restart` flag for that node.

Not blocking, can be fixed in a followup PR.

Itxaka replied:
indeed, I could do a search + filtering afterwards for the node NotFound error. Im in favor of not raising an error if the flag doesnt exists.

matelakat replied:
I modified the code to return NotFound only if the node was not found.

vuntz commented on crowbar_framework/app/controllers/api/restart_management_controller.rb:
```diff
@@ -0,0 +1,147 @@
+    proposal = Proposal.where(barclamp: barclamp).first
+
+    # validate proposal is active as we dont allow to disable restarts for first barclamp runs
+    raise Crowbar::Error::BarclampNotActive if proposal.nil?
```
I'm not convinced the check is fully fine. Let's assume we have a barclamp that supports multiple proposal. If one is active, then this will also block restart when applying a second proposal for the first time.

Itxaka replied:
+1

Itxaka commented on chef/cookbooks/utils/libraries/service.rb:
```diff
@@ -0,0 +1,79 @@
+        # if the cookbook_name is apache2, then any cookbook could have triggered it, so
+        # try to pick the value from the extra key that we set
+        cookbook = if cookbook_name == "apache2"
+          node.run_state["apache2_restart_origin"]
```
Yes, but for now the only apache2 restart comes from keystone. The rest of the cookbooks that use apache2 just do a reload, which is a better approach. Hopefully we move the keystone restarts to a reload and can drop this.

Itxaka commented on chef/cookbooks/utils/libraries/service.rb:
```diff
@@ -0,0 +1,79 @@
+        node.set[:crowbar_wall][:requires_restart] = {} \
+          unless node[:crowbar_wall][:requires_restart]
+        node.set[:crowbar_wall][:requires_restart][cookbook] = {} \
+          unless node[:crowbar_wall][:requires_restart][cookbook]
```
Yes, they are identical. the thing is, during the chef run we only know the cookbook name, not the barclamp and during the crowbar part we only know the barclamp but not the cookbook. But it should not matter as they are the same in both cases.

Itxaka commented on crowbar_framework/app/controllers/api/restart_management_controller.rb:
```diff
@@ -0,0 +1,147 @@
+  # beforehand
+  def restarts_get
+    # maybe we want to blacklist some barclamps from the restart list?
+    blacklist = []
```
+1

Itxaka commented on crowbar_framework/app/controllers/api/restart_management_controller.rb:
```diff
@@ -0,0 +1,147 @@
+    barclamps = BarclampCatalog.members("openstack").keys.reject { |m| blacklist.include? m }
+    # Chef query search to find all nodes that have requires_restart attribute not empty
+    nodes_with_restarts = NodeObject.find("requires_restart:*")
+    @final_list = {}
```
+1

Itxaka commented on chef/cookbooks/utils/libraries/service.rb:
```diff
@@ -0,0 +1,79 @@
+        end
+
+        # if the databag or item does not exits it returns a 404
+        data_bag = data_bag_item("crowbar-config", "disallow_restart") rescue false
```
+1

vuntz commented:
In case it wasn't noticed, CI failed due to a bug in the pull request.

Itxaka commented:
> I'm not a fan of /restart_management as the route name - seems overly verbose. Why not just /restarts? So then you'd have /restarts/config and /restarts/list.

I thougth about it but it seem like a very generic name. Its not helpful that the routes changed names several times <sup> and that Im terrible at naming things</su>

> This seems like a fairly major change. Is there a design document or etherpad for it somewhere which explains the impact on the various phases of the overall orchestration and the way that maintenance procedures are done? For example, how does it interact with use of Pacemaker's maintenance mode?

Not really, no. Plus it has changed a lot since inception so any specs would be really outdated by now. There is a different PR to deal with pacemaker but it basically should not affect "anything", it just skips the restart action. We leave the intended consequences to the end user. So more like an advanced unsupported feature that SAP asked for? Maybe @matelakat or @vuntz can answer that better than me, Im just a lowly grunt coder :P

> @Itxaka Thanks a lot for the helpful PR description here! Just checking - presumably you meant "Several new routes on Crowbar framework", right?

woops. Too much time dealing with chef.

aspiers commented on chef/cookbooks/utils/libraries/restart_manager.rb:
```diff
@@ -0,0 +1,52 @@
+    end
+
+    def cookbook
+      # if the cookbook_name is apache2, then any cookbook could have triggered it, so
```
I guess we'll need a similar approach for haproxy and any other shared services in the future?

vuntz replied:
I don't think so. The problem with apache is that we reload it when, say, keystone.conf is changing. This is not true for haproxy.

aspiers replied:
@vuntz I don't understand you.  Couldn't any cookbook using haproxy could trigger the need for haproxy to restart, if that cookbook changes its usage of haproxy?  Or is your point that that would be done with a reload not a restart?

vuntz replied:
The point is that haproxy is not restarted when keystone.conf/ceilometer.conf/etc. changes.

Let's take concrete examples:

- keystone.conf is changed due to, say, some ldap config change. This would usually trigger a restart/reload from apache from within the keystone cookbook.
- ceilometer.conf is changed due to whatever config change. This would usually trigger a restart/reload from apache from within the ceilometer cookbook.

and:

- haproxy.conf is changed for some change specific to the keystone backend. This would trigger a restart/reload from the crowbar-pacemaker cookbook.
- haproxy.conf is changed for some change specific to the ceilometer backend. This would trigger a restart/reload, still from the crowbar-pacemaker cookbook.

The difference is that apache can be restarted from multiple cookbooks. This is not true for haproxy where the service restart/reload only occurs from one place, even though it is true that the config for haproxy is influenced by several cookbooks.

aspiers commented on crowbar_framework/app/controllers/api/restart_management_controller.rb:
```diff
@@ -0,0 +1,168 @@
+# limitations under the License.
+#
+
+class Api::RestartManagementController < ApiController
```
I agree with the dog here.

matelakat replied:
changing that.

aspiers replied:
Not changed yet AFAICS?

Itxaka replied:
I changed it back as per the comment on the PR:

>I reverted back the Api::RestartManagementController format. I do not care what hound says in this case, all other Api classes have the same format, consistency over correctness should be ok in this case.

>I also fixed the tests. Lets pray this one is the good one.

aspiers commented on crowbar_framework/app/controllers/api/restart_management_controller.rb:
```diff
@@ -0,0 +1,148 @@
+    # validate that the barclamp exists
+    raise Crowbar::Error::NotFound unless managed_barclamps.include? barclamp
+
+    proposal = Proposal.where(barclamp: barclamp).first
```
The dog is right here too.

matelakat replied:
actually we don't even need that line, proposal is never used, that was removed as an earlier change.

aspiers commented on crowbar_framework/app/controllers/api/restart_management_controller.rb:
```diff
@@ -0,0 +1,148 @@
+    head :ok
+  end
+
+  private
```
And here.

matelakat replied:
yep, was a duplicate.

aspiers commented on crowbar_framework/spec/controllers/api/restart_management_controller_spec.rb:
```diff
@@ -0,0 +1,223 @@
+require "spec_helper"
+require "json"
+
+describe Api::RestartManagementController, type: :request do
```
IIRC this false positive from Hound was fixed recently.  Maybe a rebase would get rid of it?  And kudos for writing so much test code that it triggered the warning!! :grin:

aspiers commented:
Even though this is a fairly major architectural change which deserves more eyeballs than normal, it still seems like another case of #toomanyreviewers to me. (N.B. not a criticism of anyone here, just collecting more [evidence](https://github.com/search?q=toomanyreviewers&type=Issues) for [the need to optimise distribution of review work](http://mailman.suse.de/mlarch/SuSE/cloud-devel/2017/cloud-devel.2017.08/msg00207.html).)

matelakat commented:
@rsalevsky @aspiers I believe issues addressed, re-review?

aspiers commented:
@matelakat OK I'm persuaded on the monkey patch, but this concerns me:

>> This seems like a fairly major change. Is there a design document or etherpad for it somewhere which explains the impact on the various phases of the overall orchestration and the way that maintenance procedures are done? For example, how does it interact with use of Pacemaker's maintenance mode?
>
> Not really, no. Plus it has changed a lot since inception so any specs would be really outdated by now. There is a different PR to deal with pacemaker but it basically should not affect "anything", it just skips the restart action. We leave the intended consequences to the end user. So more like an advanced unsupported feature that SAP asked for? Maybe @matelakat or @vuntz can answer that better than me, Im just a lowly grunt coder :P

Hopefully the new maintenance workflows resulting from this have been thought out (usually @vuntz has a good picture in his head of the whole thing), but if not then personally I'm pretty sceptical that it's a good idea to merge this until they have.  Otherwise we could be missing some serious flaw in the overall design.

Itxaka commented:
Tests broken, probably due to the `API` class change, more time lost :(
```/home/travis/build/crowbar/crowbar-core/crowbar_framework/app/controllers/api/restart_management_controller.rb:17:in `<top (required)>': Api is not a class (TypeError)```

Itxaka commented:
I reverted back the `Api::RestartManagementController` format. I do not care what hound says in this case, all other Api classes have the same format, consistency over correctness should be ok in this case.

I also fixed the tests. Lets pray this one is the good one.

matelakat commented:
> Hopefully the new maintenance workflows resulting from this have been thought out (usually @vuntz has a good picture in his head of the whole thing), but if not then personally I'm pretty sceptical that it's a good idea to merge this until they have. Otherwise we could be missing some serious flaw in the overall design.

@aspiers I cannot report what is in @vuntz 's head, but he has seen this PR several times, and asked me as well to push it through. The feature itself when turned off should not have an impact on the existing behaviour, so I consider this safe to be merged. It can be a lot help when people want to prevent service re-starts.

Itxaka commented:
@aspiers re-review or dismissal of the review pls? :)

matelakat commented:
Given that @vuntz has left a comment on https://github.com/crowbar/crowbar-ha/pull/224#discussion_r140273431 I am adding a re-factor step to this change as per his request.

Itxaka commented:
Tests failing indicate that the route constraints work :D

aspiers commented on crowbar_framework/config/routes.rb:
```diff
@@ -268,5 +268,18 @@
+
+    # service restart management
+    resource :restart_management,
+      constraints: lambda { |r| Rails.application.config.experimental["disallow_restart"]["enabled"] },
```
In addition to Hound's complaints, won't this risk causing ```NoMethodError: undefined method `[]' for nil:NilClass``` since `config.experimental` could be defined but missing `disallow_restart`?

Itxaka replied:
yep, fixing this, the tests and adding more tests for the route error if the feature is disabled

Itxaka commented on crowbar_framework/config/routes.rb:
```diff
@@ -268,5 +268,20 @@
+
+    # service restart management
+    resource :restart_management,
+      constraints: ->(request) {
```
OMG

vuntz commented on chef/cookbooks/utils/libraries/restart_manager.rb:
```diff
@@ -0,0 +1,52 @@
+      # use the service as the key name for the data for easy removal
+      # force it to be a simple hash instead of a node attribute if you want the
+      # update method to actually update. If you update the attribute directly it doesn't
+      # update but overwrites! The joys of chef!
```
Can we call `add_restart_management_node_attributes` from here, instead of having the users of this lib do it? The rationale is that we would only create `@node[:crowbar_wall][:requires_restart]` if really required. And it matters because in the rails app, we do a search query for nodes with that attribute -- right now, all nodes would have that attribute, which is not great.

Itxaka replied:
Umm that is a good point and indeed, we are setting the attributes in every node for no reason.

Changing it!

vuntz commented on chef/cookbooks/utils/libraries/restart_manager.rb:
```diff
@@ -0,0 +1,52 @@
+    end
+
+    def clear_restart_requests_for(service_name)
+      @node[:crowbar_wall][:requires_restart][cookbook].delete(service_name)
```
If we do the change I just suggested, we need to be safer with something like:

```
 @node.fetch(:crowbar_wall, {}).fetch(:requires_restart, {}).fetch(cookbook, {}).delete(service_name)
```

Itxaka replied:
+1

vuntz commented on crowbar_framework/app/controllers/api/restart_management_controller.rb:
```diff
@@ -0,0 +1,144 @@
+        # skip if there is not attributes for the barclamp or the requires_restart is empty
+        # why node.node.fetch? Dont ask me, ask the chef creators why this works
+        # https://github.com/SUSE-Cloud/chef/blob/10-stable-suse/chef/lib/chef/node.rb#L187
+        requires_restart = node.node.fetch(
```
`node.node`? Note that you can use `node.crowbar_wall.fetch(...` as `crowbar_wall` is a method that does what you want.

Itxaka replied:
+1

Itxaka replied:
Interesting that changing this makes the test fail as crowbar_wall is not returning the expected values. Not sure if its due the test environment and out fake nodes, or it just has a different approach

Itxaka replied:
Disregard the above comment. I did not change the fetching correctly so of course it didnt work. Duh

vuntz commented on crowbar_framework/app/controllers/api/restart_management_controller.rb:
```diff
@@ -0,0 +1,147 @@
+        ).fetch(
+          "requires_restart", {}
+        ).fetch(
+           barclamp.to_s, {}
```
Hound is correct here.

vuntz replied:
Btw, it's worth mentioning one thing here (and I would add a comment for this): you rely on the fact that the barclamp name matches the cookbook name. This is generally true, but not always. But for what we're interested in, it is true. It's okay to leave this like this for now, especially with the experimental flag.

Actually, half-changing my mind here: would it make sense to rename everything from `barclamp` to `cookbook` in this code? You would have `managed_cookbooks`, and this is where you would have a comment saying "right now, we only look at openstack, and the cookbook names are the same as the barclamp names, so we can use the barclamp catalog".

I'll leave the final call to you.

Itxaka replied:
Sounds good, and maybe it would be clearer to newcomers to the patch. And Im all in for it being clearer, so changing!

Itxaka commented:
@vuntz updated!

Itxaka commented:
ha job is failing due to an unrelated change to automation, fix at https://github.com/SUSE-Cloud/automation/pull/2236

vuntz commented on chef/cookbooks/utils/libraries/service.rb:
```diff
@@ -0,0 +1,47 @@
+        if restart_manager.disallow_restart?
+          Chef::Log.info("Disallowing restart for #{@new_resource.name} due to flag")
+          restart_manager.register_restart_request(@new_resource.name.to_sym, false)
+
```
Why an empty line? :-)

Itxaka replied:
ᕙ(⇀‸↼‶)ᕗ

matelakat commented on chef/cookbooks/utils/libraries/service.rb:
```diff
@@ -0,0 +1,47 @@
+
+        if restart_manager.disallow_restart?
+          Chef::Log.info("Disallowing restart for #{@new_resource.name} due to flag")
+          restart_manager.register_restart_request(@new_resource.name.to_sym, false)
```
I noticed that we have moved away from creating the RestartManager with the service name, and instead we are adding it as a parameter for each of the methods - why didn't we go with specifying the service name for the constructor?

Itxaka replied:
Im not sure when did that happen, I dont think I changed that, but I can see that on the crowbar-pacemaker the signature for the RestartManager is different :/

Maybe I magled it somehow, let me re-fix that :(

Itxaka commented:
@vuntz @matelakat updated, changes for easy re-review:
 - RestartManager is now initialized with 4 params instead of 2 (cookbook_name, node, new_resource, is_pacemaker_service)
 - Due to the change above, callnig the rest of the RestartManager methods do not require to pass any arguments as they are already in the object
 - Removed empty line :P

Itxaka commented:
> Assuming only chef/cookbooks/utils/libraries/restart_manager.rb got changed.

service.rb also got changed to remove the params not needed for `RestartManager` and to remove an empty line

Itxaka commented:
pushed to fix a small typo: `clear_restart_request` -> `clear_restart_requests`

Itxaka commented:
>I think I'm OK with this now, assuming the use of _ vs. - in databags is consistent with whatever's already there.

```
root@crowbar:~ # knife data bag list
  crowbar
  crowbar-config
  repos-suse-12_2-aarch64
  repos-suse-12_2-s390x
  repos-suse-12_2-x86_64
root@crowbar:~ # knife data bag show crowbar-config
  core
  openstack
root@crowbar:~ # knife data bag show crowbar
  admin_network
  bmc_network
  bmc_vlan_network
  nova_fixed_network
  nova_floating_network
  os_sdn_network
  public_network
  storage_network
root@crowbar:~ # knife data bag show repos-suse-12_2-x86_64
  SLES12-SP2-Updates-test
  cloud
  ptf
  sles12-sp2-pool
  sles12-sp2-updates
```

as you can see databag **names**  and databag **items** are a mix of `_` and `-`

tbh I dont care either way, unless there is a wide agreement on how to implement this going forward although I would say that it doesnt really matter as it depends on what you are saving to the databag as long as it mimics whats in the code, i.e. saving admin_network attributes into the databags should go to admin_network, **not** admin-network even if thats the agreed way forward. So logic instead of hard rules. But thats my opinion only and believes that logic is the same for everyone. Yeah right :P

Plus I think that after almost 300 comments, several code revisions, github dying on this PR page load (lol) and a really short timeframe to finally get this in time for SAP deployment, the databag item format should be the least of our concerns, as we can change it afterwards with no impact (no migrations needed for that type of change), so that discussion should not be a blocker for this PR 👍

aspiers commented:
> as you can see databag names and databag items are a mix of _ and -
>
> tbh I dont care either way, unless there is a wide agreement on how to implement this going forward although I would say that it doesnt really matter as it depends on what you are saving to the databag as long as it mimics whats in the code, i.e. saving admin_network attributes into the databags should go to admin_network, not admin-network even if thats the agreed way forward. So logic instead of hard rules. But thats my opinion only and believes that logic is the same for everyone. Yeah right :P

I realise it's a mix. My point was that if it's 80% one way and 20% the other, then the stuff we're adding should go with the 80%.  But yeah, I'm as fed up with this PR as you by now, so I'm approving ;-)

Itxaka commented:
![parrot](https://user-images.githubusercontent.com/1447686/31408175-f59449f6-ae07-11e7-96a0-0345f09447b3.gif)
